File: InstanceSide/BallPassSchedulePattern/Players_AiTypes/AffinePlayerDir/AffineLayer.cs
using UdonSharp;
using UnityEngine;

public class AffineLayer : UdonSharpBehaviour
{
    // RinaNumpyのアタッチ
    public RinaNumpy rNp;

    // クラス内で保持する変数
    private float[][] weights; // 重み
    private float[] bias; // バイアス
    private float[] xArray; // 入力
    private float[] yArray; // 出力

    // 初期化メソッド
    public void Initialize(float[][] initialWeights, float[] initialBias)
    {
        // 重みとバイアスを初期化
        weights = initialWeights;
        bias = initialBias;
    }

    public float[] Forward(float[] input)
    {
        // 入力を保持
        xArray = input;

        // Affine変換: Wx + b をRinaNumpyを使って計算
        yArray = rNp.Add_FloatArray_FloatArray(
            rNp.DotProduct_FloatArray2D_FloatArray(weights, xArray),
            bias
        );

        // 出力を返す
        return yArray;
    }

    public float[] Backward(float[] dout)
    {
        // 重みの勾配dWを計算
        float[][] dW = new float[weights.Length][];
        for (int i = 0; i < weights.Length; i++)
        {
            // RinaNumpyを活用して、dout[i]とxArrayを掛け算
            dW[i] = rNp.Multiply_FloatArray_Float(xArray, dout[i]);
        }

        // バイアスの勾配dbを計算
        float[] db = rNp.Copy_FloatArray(dout); // doutをそのままコピー

        // 入力に対する誤差dxを計算
        float[] dx = rNp.DotProduct_FloatArray2D_FloatArray(
            rNp.TransposeMatrix(weights),
            dout
        );

        // 必要であれば、ここでdWとdbを別の処理に渡す
        // 例: 学習用のOptimizerに渡して更新する

        // 計算されたdxを返す
        return dx;
    }
}

File: ServerSide/BallPassSchedulePattern/Players_AiTypes/AffinePlayerDir/AffineLayer.py
import numpy as np

class AffineLayer:
    def __init__(self, initial_weights, initial_bias):
        """
        初期化
        :param initial_weights: 重み (2DリストまたはNumPy配列)
        :param initial_bias: バイアス (1DリストまたはNumPy配列)
        """
        self.weights = np.array(initial_weights)  # 重み
        self.bias = np.array(initial_bias)  # バイアス
        self.x_array = None  # 入力
        self.y_array = None  # 出力

    def forward(self, input_data):
        """
        Forward処理: Wx + b を計算
        :param input_data: 入力 (1DリストまたはNumPy配列)
        :return: 出力 (1D NumPy配列)
        """
        self.x_array = np.array(input_data)
        self.y_array = np.dot(self.weights, self.x_array) + self.bias
        return self.y_array

    def backward(self, dout):
        """
        Backward処理: 勾配の計算
        :param dout: 出力側からの誤差勾配 (1DリストまたはNumPy配列)
        :return: dx (入力側への誤差勾配)
        """
        dout = np.array(dout)
        
        # 重みの勾配 (dW)
        dW = np.outer(dout, self.x_array)  # 各doutに対してxArrayを掛ける

        # バイアスの勾配 (db)
        db = np.copy(dout)  # doutそのままコピー

        # 入力に対する誤差 (dx)
        dx = np.dot(self.weights.T, dout)  # 重み行列の転置を使用

        # 必要に応じてdW, dbを外部で利用可能
        self.dW = dW
        self.db = db

        return dx
